Make gpinitsystem use walrep mirrors instead of filerep mirrors

jimmyyih opened the pull request:
gpinitsystem with walrep mirrors instead of filerep mirrors

With file replication gone, gpinitsystem should no longer try to
initialize the cluster through filerep sequence.

The sequence now goes as follows:
1. Create and start master in master-only mode
2. Create primaries and register to master
3. Stop master.
4. Run gpstart to start master and primaries.
5. Create mirrors w/ pg_basebackup and register to master.
6. Start the mirrors and wait until primaries and mirrors sync.

xinzweb commented on concourse/pipelines/pipeline_segwalrep.yml:
```diff
@@ -1806,3 +1836,172 @@ jobs:
+      passed: [gate_mm_start]
+      trigger: true
+    - get: centos-gpdb-dev-6
+##  - aggregate: *gets_for_behave
```
We don't need it? then, maybe remove it.

xinzweb commented on concourse/pipelines/pipeline_segwalrep.yml:
```diff
@@ -1806,3 +1836,172 @@ jobs:
+    params:
+      BEHAVE_TAGS: gpinitsystem
+
+## - name: MM_pt-rebuild
```
same

xinzweb commented on concourse/pipelines/pipeline_segwalrep.yml:
```diff
@@ -1806,3 +1836,172 @@ jobs:
+##         TEST_OS: centos
+##         CONFIGURE_FLAGS: {{configure_flags}}
+
+## - name: MU_check_centos
```
same

xinzweb commented on concourse/pipelines/pipeline_segwalrep.yml:
```diff
@@ -1806,3 +1836,172 @@ jobs:
+    - get: gpdb_src
+      passed:
+      - MM_gpinitsystem
+##      - MM_gpcheck
```
same

xinzweb commented on concourse/pipelines/templates/segwalrep-tpl.yml:
```diff
@@ -16,14 +16,19 @@
-{% set MM_gpexpand_suites = ['gpexpand_1',
-                             'gpexpand_2'] %}
+{% set MM_1_suites = [] %}
+## Add these back to MM_1_suites list if needed later
```
why do we want to remove from the template? We shouldn't touch the template, right?

jimmyyih replied:
The template controls the pipeline file.  You make changes here and run the pipeline generator script.

xinzweb commented on gpAux/gpdemo/Makefile:
```diff
@@ -45,12 +45,7 @@ all:
-	./gpsegwalrep.py init --host `hostname`; \
-	./gpsegwalrep.py start;
-endif
+	@./demo_cluster.sh
```
Nice!

xinzweb commented on gpMgmt/bin/gpinitsystem:
```diff
@@ -1300,7 +1300,7 @@ CREATE_QD_DB () {
 			ERROR_EXIT "[FATAL]:-Could not establish connection to hostname $GP_HOSTNAME. Please check your configuration." 2
 		fi
-		UPDATE_GPCONFIG $GP_PORT $GP_DBID $GP_CONTENT $GP_HOSTNAME $GP_HOSTADDRESS $GP_PORT $GP_DIR p $GP_REPLICATION_PORT
+		UPDATE_GPCONFIG $GP_PORT $GP_DBID $GP_CONTENT $GP_HOSTNAME $GP_HOSTADDRESS $GP_PORT $GP_DIR p
```
Why a tailing `p` there? Is that possible to remove it by adjust the script? Is that something relates to the master standby setup logic later on?

jimmyyih replied:
Tailing p is to signify primary role I think.  It can probably be removed but we'll be leaving it as is because gpinitsystem will be rewritten and we just need this utility working.

xinzweb commented on gpMgmt/bin/gpinitsystem:
```diff
@@ -1373,27 +1365,9 @@ LOAD_QE_SYSTEM_DATA () {
 					ERROR_EXIT "[FATAL]:-Could not establish connection to hostname $GP_HOSTNAME. Please check your configuration." 2
 				fi				
-				UPDATE_GPCONFIG $MASTER_PORT $GP_DBID $GP_CONTENT $GP_HOSTNAME $GP_HOSTADDRESS $GP_PORT $GP_DIR p $GP_REPLICATION_PORT
+				UPDATE_GPCONFIG $MASTER_PORT $GP_DBID $GP_CONTENT $GP_HOSTNAME $GP_HOSTADDRESS $GP_PORT $GP_DIR p
```
similar q.

xinzweb commented on gpMgmt/bin/gpinitsystem:
```diff
@@ -1625,6 +1599,25 @@ CREATE_DATABASE () {
+        if [ x"$RESULT" == x"f" ]; then
+            break
+        fi
+        sleep 2;
```
Can we reduce to 1 second, or even faster like 0.1 second? We can loop more times, but we should get gpinitsystem to finish as fast as possible. Also, why give it 5min timeout? Do we really wait for 5min? I feel way less patient on waiting for 5min. This is already after basbackup, and system should be up and running already. Maybe a 10 seconds total wait time would be good enough?

jimmyyih replied:
Sure, we can loop faster.  The 5 minute timeout is mostly there because Marbin and I could not decide on a good number.  Is a 1 minute timeout a good compromise for you?

xinzweb commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -137,88 +137,96 @@ PROCESS_QE () {
-    PARA_EXIT $RETVAL "Update checkpoint segments"
+    # do primary stuff
+    if [ x"" = x"$COPY_FROM_PRIMARY_HOSTADDRESS" ]; then
+        # Configure postgresql.conf
```
Is this whole block better to be refactored as another function? Seems like the only diff caused by the indent of the code? something like INIT_PRIMARY()

jimmyyih replied:
Same as above, gpinitsystem will be rewritten instead of refactored.

xinzweb commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -137,88 +137,96 @@ PROCESS_QE () {
     if [ x"" != x"$COPY_FROM_PRIMARY_HOSTADDRESS" ]; then
-        LOG_MSG "[INFO]:-Copying data for mirror on ${GP_HOSTADDRESS} using remote copy from primary ${COPY_FROM_PRIMARY_HOSTADDRESS} ..." 1
-        RUN_COMMAND_REMOTE ${COPY_FROM_PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; ${GPHOME}/bin/lib/pysync.py -x pg_log -x postgresql.conf -x postmaster.pid ${COPY_FROM_PRIMARY_DIR} \[${GP_HOSTADDRESS}\]:${GP_DIR}"
+        LOG_MSG "[INFO]:-Running pg_basebackup to init mirror on ${GP_HOSTADDRESS} using primary on ${COPY_FROM_PRIMARY_HOSTADDRESS} ..." 1
```
Can this block be refactored to something like INIT_MIRROR()?

jimmyyih replied:
It can be refactored but will not be done in this PR because gpinitsystem will be rewritten (possibly in golang) before 6.0 is cut.

jchampio commented on concourse/pipelines/pipeline_segwalrep.yml:
```diff
@@ -1806,3 +1836,172 @@ jobs:
+          repository: alpine
+      run:
+        path: 'sh'
+        args: ['-c', 'sleep `expr 3 \* {{group_start_delay}} + 1 + $RANDOM % {{group_delay_jitter}}`']
```
To clarify: it's the template changes, not the pipeline.yml changes, that need the most in-depth review. Many of the YAML "changes" are introduced because of code that wasn't actually changed here -- this is one example.

jimmyyih commented:
Pipeline stuff has been removed from this PR because https://github.com/greenplum-db/gpdb/commit/2bbe0ada167c4148888096993171cd8a53b5802e has already been pushed in to enable all tests.

@xinzweb @asimrp I changed the force fts probe function to loop 1 second with 60 retries, moved gp_segment_configuration check before the probe, and error out after all retries.

jimmyyih commented:
Merged into walreplication branch:
https://github.com/greenplum-db/gpdb/commit/07047fa98b2c010d14875f9a67afef29984304b3
